dep0/lib/03_typecheck/src/beta_delta_reduction.cpp

```diff
@@ -68,28 +93,425 @@ bool beta_delta_normalize(environment_t const& env, func_def_t& def)
     }
     changed |= beta_delta_normalize(env, ctx, def.value.ret_type.get());
     changed |= beta_delta_normalize(env, ctx, def.value.body);
+    changed |= beta_delta_normalize(env, ctx, def.properties.sort.get());
     return changed;
 }
 
 bool beta_delta_normalize(environment_t const& env, context_t const& ctx, body_t& body)
 {
-    bool changed = beta_normalize(body);
-    while (delta_reduce(env, ctx, body))
+    auto const is_return = [] (stmt_t const& x)
+    {
+        return std::holds_alternative<stmt_t::return_t>(x.value);
+    };
+    // NB taking `stmts` by value because we are about to perform a destructive self-assignment
+    auto replace_with = [&] (std::vector<stmt_t>::iterator it, std::vector<stmt_t> stmts)
+    {
+        // If there is nothing to replace `it` with, then just remove it;
+        // otherwise we need to splice all the replacing statements in place of `it`,
+        // which can be done efficiently by replacing `it` with the first statement, and insert the rest after it.
+        // Finally, if the new statements contain a return statment, any other statement after that can be dropped.
+        if (stmts.empty())
+            return body.stmts.erase(it);
+        auto const index_of_ret = [&]
+        {
+            auto const j = std::ranges::find_if(stmts, is_return);
+            return j == stmts.end() ? std::nullopt : std::optional{std::distance(stmts.begin(), j)};
+        }();
+        auto j = stmts.begin();
+        *it++ = std::move(*j++);
+        it = body.stmts.insert(it, std::make_move_iterator(j), std::make_move_iterator(stmts.end()));
+        if (index_of_ret)
+            // if you were expecting a `+1` here, the reason it's missing is because we manually incremented `it` once
+            body.stmts.erase(it + *index_of_ret, body.stmts.end());
+        return it;
+    };
+    bool changed = false;
+    auto it = body.stmts.begin();
+    while (it != body.stmts.end())
+    {
+        changed |= beta_delta_normalize(env, ctx, *it);
+        it = match(
+            it->value,
+            [&] (expr_t::app_t& app)
+            {
+                // if immediately-invoked lambda we can extract the body,
+                // but have to suppress return statements from the invoked lambda,
+                // otherwise they would become early returns from the parent function
+                // TODO blindly suppressing return of the invoked lambda will be wrong once we have side-effects
+                if (app.args.empty())
+                    if (auto* const abs = std::get_if<expr_t::abs_t>(&app.func.get().value))
+                    {
+                        changed = true;
+                        // TODO fix this: if body contains 2 return statements, all others in between must be dropped
```

Comment:
Should be easy, just do a first pass over the body removing all statements after the 1st return. Then run the existing logic. Test code could be this
```
i32_t f()
{
    return 1;
    return 2;
}

i32_t g()
{
    return f();
}
```